### ISSUE TITLE
Multiple tests are failing with upcoming OpenSSL 3.0 release

### ISSUE THREAD

**Issue by ismail** [qpdf/qpdf#443]
Hi,

We are doing a test run with upcoming OpenSSL 3.0 release (we are testing 3.0 alpha2 atm.) and multiple tests are failing, e.g:

> cwd: /home/abuild/rpmbuild/BUILD/qpdf-10.0.1/qpdf/qtest/qpdf
> command: qpdf-ctest 2 enc-R2,V1,U=view,O=view.pdf '' a.qdf
> expected output in c-invalid-password.out
>  at qpdf.test line 3503.
> --> BEGIN EXPECTED OUTPUT <--
> error: enc-R2,V1,U=view,O=view.pdf: invalid password
>   code: 4
>   file: enc-R2,V1,U=view,O=view.pdf
>   pos : 0
>   text: invalid password
> --> END EXPECTED OUTPUT <--
> --> BEGIN ACTUAL OUTPUT <--
> error: openssl error
>   code: 2
>   file: 
>   pos : 0
>   text: openssl error
> --> END ACTUAL OUTPUT <--
> --> DIFF EXPECTED ACTUAL <--
> --- /tmp/testtemp.12117/expected	2020-05-25 08:16:22.300000000 +0000
> +++ /tmp/testtemp.12117/actual	2020-05-25 08:16:22.300000000 +0000
> @@ -1,5 +1,5 @@
> -error: enc-R2,V1,U=view,O=view.pdf: invalid password
> -  code: 4
> -  file: enc-R2,V1,U=view,O=view.pdf
> +error: openssl error
> +  code: 2
> +  file: 
>    pos : 0
> -  text: invalid password
> +  text: openssl error
> 
> warning: c-r2.pdf: object 15/0: error reading object: openssl error
>   code: 5
>   file: c-r2.pdf
>   pos : 0
>   text: object 15/0: error reading object: openssl error
> error: openssl error
>   code: 2
>   file: 
>   pos : 0
>   text: openssl error
> --> END ACTUAL OUTPUT <--
> 
> +warning: c-r2.pdf: object 15/0: error reading object: openssl error
> +  code: 5
> +  file: c-r2.pdf
> +  pos : 0
> +  text: object 15/0: error reading object: openssl error
> +error: openssl error
> +  code: 2
> +  file: 
> +  pos : 0
> +  text: openssl error
> --> END DIFFERENCES <--
> 
> warning: c-r3.pdf: object 15/0: error reading object: openssl error
>   code: 5
>   file: c-r3.pdf
>   pos : 0
>   text: object 15/0: error reading object: openssl error
> error: openssl error
>   code: 2
>   file: 
>   pos : 0
>   text: openssl error
> --> END ACTUAL OUTPUT <--
> 
> +warning: c-r3.pdf: object 15/0: error reading object: openssl error
> +  code: 5
> +  file: c-r3.pdf
> +  pos : 0
> +  text: object 15/0: error reading object: openssl error
> +error: openssl error
> +  code: 2
> +  file: 
> +  pos : 0
> +  text: openssl error
> --> END DIFFERENCES <--
> 

I'm attaching the full build log for reference.

[opeensl3-log.txt.zip](https://github.com/qpdf/qpdf/files/4680023/opeensl3-log.txt.zip)

**Comment by ismail** [comment]
I'm not sure how these test PDFs are encrypted, but I am guessing they are using a legacy method, and we are hit by https://github.com/openssl/openssl/issues/9349

**Comment by jberkenbilt** [comment]
@ismail The amount of actual openssl code in qpdf is pretty small and is all contained within a single source file, [libqpdf/QPDFCrypto_openssl.cc](https://github.com/qpdf/qpdf/blob/master/libqpdf/QPDFCrypto_openssl.cc). I would be interested in ensuring that qpdf works with openssl 3, and if possible, also with older versions. I would welcome a patch or pull request, or even just a quick pointer saying, "Use this instead of that." I can't imagine this would be hard to fix. The openssl code was contributed by another user. I didn't write it myself. But it is a full part of qpdf, and I would like it to be done correctly and in a way that will work with whatever supported version of openssl someone may be using. Thanks.

**Comment by ismail** [comment]
Yes, this is ok. Once OpenSSL comes up with a fixed API, I or someone else can easily fix this. I just wanted the problem to be tracked :-)

**Comment by p00ya** [comment]
From https://www.openssl.org/docs/OpenSSLStrategicArchitecture.html it looks like OpenSSL 3 is a large departure from OpenSSL 1.1.0 and not intended to be API compatible.  While we could hack it in to the current qpdf crypto-openssl provider with some autoconf checks and preprocessor guards, I think it would be cleaner to support from a new qpdf crypto provider.  There are no doubt parts we could factor out and share between the two.

A second reason for making a new provider is that currently OpenSSL 1.1.0 and BoringSSL are substitutable so far as qpdf is concerned.  It doesn't look like BoringSSL will have a 3.0-equivalent API.

Finally, OpenSSL 3.0.0 still isn't released, so it feels premature for qpdf to support it.  However, a continuous integration test (marked as an expected failure) could be useful in the meantime to help track known failures.

**Comment by jberkenbilt** [comment]
I have no problem with having a completely separate provider for openssl 3 and am inclined to agree that that is the correct approach. I don't care whether we factor out common sections or not. It might complicate the build a little, and there's very little code in the crypto provider.

**Comment by ismail** [comment]
No one is rushing to support openssl3 yet ;) Anyhow OpenSSL has version macros which we can use. I'm just waiting for them to release a stable before working on this.

tl;dr It's not my intention to break what's working

**Comment by jberkenbilt** [comment]
@ismail Current master now works with both openssl 1 and 3. The windows builds for 10.4 will be built with openssl 3. The detection is made at compile time. Basically the issue was that rc4 and md5, both of which are unavoidable in PDF, are in the legacy provider, so I had to change the code to load those, plus update for a few other minor API issues. It took me about 2 or 3 hours to get everything working and tested, most of which was getting it to work on Windows.
